File: fractal_sequences/pipeline.py
"""High-level pipeline for reproducible sequence explorations."""

from __future__ import annotations

from .analyzer import analyze_recurrence
from .config import AnalysisConfig, DiscretizationConfig, ExplorerConfig, RecurrenceConfig
from .coordinates import (
    AngleProjectionCoordinates,
    ComplexCartesianCoordinates,
    ComplexPolarCoordinates,
    CoordinateSystem,
    ProjectionXCoordinates,
    ProjectionYCoordinates,
    RadiusProjectionCoordinates,
)
from .discretization import DiscretizationGrid
from .examples import cosine_fixed_point, logistic_map, quadratic_complex_map
from .recurrence import Recurrence
from .transformations import DiscreteAnalysisResult, DiscreteTransformer


def run_exploration(config: ExplorerConfig) -> DiscreteAnalysisResult:
    """Run recurrence analysis and discrete transformation from one config."""

    recurrence = build_recurrence(config.recurrence)
    coordinate_system = build_coordinate_system(config.discretization.coordinate_system)
    grid = build_grid(config.discretization)
    transformer = DiscreteTransformer(coordinate_system, grid)
    analysis = config.analysis

    result = analyze_recurrence(
        recurrence,
        max_terms=analysis.max_terms,
        tolerance=analysis.tolerance,
        divergence_limit=analysis.divergence_limit,
        attractor_window=_attractor_window(analysis),
        max_period=analysis.max_period,
        min_repetitions=analysis.min_repetitions,
        transient_terms=analysis.transient_terms,
        stability_ratio=analysis.stability_ratio,
        cluster_tolerance=analysis.cluster_tolerance,
    )

    return DiscreteAnalysisResult.from_analysis(
        sequence_result=result,
        transformer=transformer,
        config=config,
    )


def build_recurrence(config: RecurrenceConfig) -> Recurrence:
    params = config.parameters
    if config.name == "cosine":
        return cosine_fixed_point(x0=params.get("x0", 1.0))
    if config.name == "logistic":
        return logistic_map(
            r=params.get("r", 3.2),
            x0=params.get("x0", 0.2),
        )
    if config.name == "quadratic_complex":
        c = complex(params.get("c_real", -0.123), params.get("c_imag", 0.745))
        z0 = complex(params.get("z0_real", 0.0), params.get("z0_imag", 0.0))
        return quadratic_complex_map(c=c, z0=z0)
    raise ValueError(f"unknown recurrence config: {config.name}")


def build_coordinate_system(name: str) -> CoordinateSystem:
    if name == "x":
        return ProjectionXCoordinates()
    if name == "cartesian":
        return ComplexCartesianCoordinates()
    if name == "polar":
        return ComplexPolarCoordinates()
    if name == "y":
        return ProjectionYCoordinates()
    if name == "radius":
        return RadiusProjectionCoordinates()
    if name == "angle":
        return AngleProjectionCoordinates()
    raise ValueError(f"unknown coordinate system config: {name}")


def build_grid(config: DiscretizationConfig) -> DiscretizationGrid:
    coordinate_system = build_coordinate_system(config.coordinate_system)
    missing_axes = set(coordinate_system.axes) - set(config.ranges)
    if missing_axes:
        missing = ", ".join(sorted(missing_axes))
        raise ValueError(f"missing discretization ranges for axes: {missing}")

    return DiscretizationGrid(
        {
            axis: config.ranges[axis].to_subdivision()
            for axis in coordinate_system.axes
        }
    )


def _attractor_window(config: AnalysisConfig) -> int:
    if config.attractor_window is not None:
        return config.attractor_window
    return max(50, config.max_terms // 3)

File: src/pipeline.rs
use std::sync::Arc;

use anyhow::{Result, bail};
use indexmap::IndexMap;
use num_complex::Complex64;

use crate::analyzer::analyze_recurrence;
use crate::config::{AnalysisConfig, DiscretizationConfig, ExplorerConfig, RecurrenceConfig};
use crate::coordinates::{
    AngleProjectionCoordinates, ComplexCartesianCoordinates, ComplexPolarCoordinates,
    CoordinateSystem, ProjectionXCoordinates, ProjectionYCoordinates, RadiusProjectionCoordinates,
};
use crate::discretization::DiscretizationGrid;
use crate::examples::{cosine_fixed_point, logistic_map, quadratic_complex_map};
use crate::recurrence::Recurrence;
use crate::transformations::{DiscreteAnalysisResult, DiscreteTransformer};

pub fn run_exploration(config: &ExplorerConfig) -> Result<DiscreteAnalysisResult> {
    let recurrence = build_recurrence(&config.recurrence)?;
    let coordinate_system = build_coordinate_system(&config.discretization.coordinate_system)?;
    let grid = build_grid(&config.discretization)?;
    let transformer = DiscreteTransformer::new(coordinate_system, grid)?;
    let analysis = &config.analysis;
    let result = analyze_recurrence(
        &recurrence,
        analysis.max_terms,
        analysis.tolerance,
        analysis.divergence_limit,
        attractor_window(analysis),
        analysis.max_period,
        analysis.min_repetitions,
        analysis.transient_terms,
        analysis.stability_ratio,
        analysis.cluster_tolerance,
    )?;
    DiscreteAnalysisResult::from_analysis(result, transformer, Some(config.clone()))
}

pub fn build_recurrence(config: &RecurrenceConfig) -> Result<Recurrence> {
    let params = &config.parameters;
    match config.name.as_str() {
        "cosine" => Ok(cosine_fixed_point(*params.get("x0").unwrap_or(&1.0))),
        "logistic" => Ok(logistic_map(
            *params.get("r").unwrap_or(&3.2),
            *params.get("x0").unwrap_or(&0.2),
        )),
        "quadratic_complex" => {
            let c = Complex64::new(
                *params.get("c_real").unwrap_or(&-0.123),
                *params.get("c_imag").unwrap_or(&0.745),
            );
            let z0 = Complex64::new(
                *params.get("z0_real").unwrap_or(&0.0),
                *params.get("z0_imag").unwrap_or(&0.0),
            );
            Ok(quadratic_complex_map(c, z0))
        }
        _ => bail!("unknown recurrence config: {}", config.name),
    }
}

pub fn build_coordinate_system(name: &str) -> Result<Arc<dyn CoordinateSystem>> {
    match name {
        "x" => Ok(Arc::new(ProjectionXCoordinates)),
        "cartesian" => Ok(Arc::new(ComplexCartesianCoordinates)),
        "polar" => Ok(Arc::new(ComplexPolarCoordinates)),
        "y" => Ok(Arc::new(ProjectionYCoordinates)),
        "radius" => Ok(Arc::new(RadiusProjectionCoordinates)),
        "angle" => Ok(Arc::new(AngleProjectionCoordinates)),
        _ => bail!("unknown coordinate system config: {name}"),
    }
}

pub fn build_grid(config: &DiscretizationConfig) -> Result<DiscretizationGrid> {
    let coordinate_system = build_coordinate_system(&config.coordinate_system)?;
    let required_axes = coordinate_system.axes();
    for axis in required_axes {
        if !config.ranges.contains_key(*axis) {
            bail!("missing discretization ranges for axes: {axis}");
        }
    }
    let mut subdivisions = IndexMap::new();
    for axis in required_axes {
        let range_config = config
            .ranges
            .get(*axis)
            .ok_or_else(|| anyhow::anyhow!("missing discretization range: {axis}"))?;
        subdivisions.insert(String::from(*axis), range_config.to_subdivision()?);
    }
    DiscretizationGrid::new(subdivisions)
}

fn attractor_window(config: &AnalysisConfig) -> usize {
    config.attractor_window.unwrap_or((config.max_terms / 3).max(50))
}
